src/softmax.jl

```diff
@@ -95,38 +105,29 @@ It is semantically equivalent to the following:
 
 See also [`softmax`](@ref).
 """
-logsoftmax(x; dims = 1) = logsoftmax!(similar(x, (float ∘ eltype)(x)), x; dims = dims)
+logsoftmax(x::AbstractArray{T}; dims = 1) where {T} = logsoftmax!(similar(x, float(T)), x; dims)
 
-logsoftmax!(x; dims = 1) = logsoftmax!(x, x; dims = dims)
+logsoftmax!(x::AbstractArray; dims = 1) = logsoftmax!(x, x; dims)
 
 function logsoftmax!(out::AbstractArray{T}, x::AbstractArray; dims = 1) where {T}
-    max_ = maximum(x; dims = dims)
+    max_ = maximum(x; dims)
     if all(isfinite, max_)
         out .= x .- max_
     else
         @. out = ifelse(isequal(max_,Inf), ifelse(isequal(x,Inf), 0, -Inf), x - max_)
     end
-    log_ = log.(sum(exp, out; dims = dims))
+    @fastmath log_ = log.(sum(exp, out; dims))
```

Comment:
This `@fastmath` is worth about a factor of 2:
```
julia> @btime softmax(x)   setup=(x=rand(Float32, 10, 1000));
  min 102.041 μs, mean 110.310 μs (6 allocations, 47.30 KiB)
  min 60.375 μs, mean 69.655 μs (6 allocations, 47.30 KiB)

julia> @btime logsoftmax(x)   setup=(x=rand(Float32, 10, 1000));
  min 126.166 μs, mean 132.883 μs (7 allocations, 51.36 KiB)
  min 67.875 μs, mean 74.167 μs (7 allocations, 51.36 KiB)
```
Maybe it's worth it? Is there an obvious accuracy check we should try out?

It shouldn't change `NaN` behaviour, since `maximum` will see any, and propagate them.

Reply:
I had a look through the historical issues around `softmax` and couldn't find anything not already covered by the test suite.

Reply:
Actually, the integration test errors in https://github.com/FluxML/Tracker.jl/blob/master/test/tracker.jl#L35 and https://github.com/FluxML/Tracker.jl/blob/master/test/tracker.jl#L38 look to be new.

Reply:
I thought those were failing on an earlier commit, without fastmath. 

Now I made https://github.com/FluxML/Tracker.jl/pull/127, which locally passes with or without this PR. We should merge that, then check that the downstream tests here pass with it, I think.